keps/sig-apps/0012-20180522-deployment-inplace-update.md

```diff
@@ -0,0 +1,215 @@
+---
+kep-number: 12
+title: Support deployment controller in-place update
+authors:
+  - "@jian-he"
+owning-sig: sig-apps
+reviewers:
+  - @janetkuo, @kargakis, @smarterclayton
+approvers:
+  - TBD
+editor: "@jian-he"
+creation-date: "2018-05-23"
+last-updated: "2018-05-31"
+status: provisional
+---
+
+# Support deployment controller in-place update
+
+Table of Contents
+=================
+
+   * [Support deployment controller in-place update](#support-deployment-controller-in-place-update)
+   * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+      * [Goals](#goals)
+      * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [API Change](#api-change)
+         * [Deployment Controller](#deployment-controller)
+            * [Update](#update)
+            * [Rollback](#rollback)
+         * [Alternatives](#alternatives)
+            * [Alternative 1](#alternative-1)
+            * [Alternative 2](#alternative-2)
+      * [Graduation Criteria](#graduation-criteria)
+            
+## Summary
+We propose a new update strategy for deployment controller to do rolling update images and/or commands of containers without tearing down the pods, i.e. in-place update. 
+
+
+## Motivation
+Today, Deployment controller, while doing rolling update, will teardown all the pods and then recreate all the pods. In essence, it reshuffles all the pods.
+This approach works in most cases but has several limitations:
+* we have thousands of pods running super critical user-facing online services with high availability guarantee. Re-shuffling all the pods over thousands of machines to do the upgrade is problematic:
+    1) Since pods are reshuffled, network topology for all the pods become indeterministic every time we do upgrade and this can cause inconsistent network performance.
+    It may also cause inconsistent runtime performance due to host hardware are different. This breaks user experience.
+    2) Certain container images are very large, re-shuffling all the pods means re-downloading the images onto the hosts if the image is not present. 
```

Comment:
The scheduler attempts to favor placement on nodes where the image is already present. Have you analyzed why that is not effective?